sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala

```diff
@@ -176,7 +176,8 @@ object TableOutputResolver extends SQLConfHelper with Logging {
     if (canWriteValue) {
       val nullCheckedValue = checkNullability(value, attr, conf, colPath)
       val casted = cast(nullCheckedValue, attrTypeWithoutCharVarchar, conf, colPath.quoted)
-      val exprWithStrLenCheck = if (conf.charVarcharAsString || !attrTypeHasCharVarchar) {
+      val exprWithStrLenCheck = if ((conf.charVarcharAsString &&
+        !conf.preserveCharVarcharTypeInfo) || !attrTypeHasCharVarchar) {
```

Comment:
why? no matter we preserver char/varchar type info or not, we should always apply length check and padding.

Reply:
The added condition was indeed unnecessary.